### PR TITLE
chore(STONEO11Y-101): Alerting rule for Quota Exceeded

### PR DESCRIPTION
**Pull request by ShakedAviv50313**
Create an alerting rule for Quota Exceeded error and add related unit test cases.

### CODE REVIEW

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 1]
```diff
@@ -25,3 +25,14 @@ spec:
       annotations:
         summary: "Pod {{ $labels.pod }} is crash looping"
         description: "Pod {{ $labels.namespace }}/{{ $labels.pod }} ({{ $labels.container }}) is in waiting state (reason: 'CrashLoopBackOff') for more than 15 minutes."
+    - alert: QuotaExceeded
+      expr: |
+        kube_resourcequota{type="used", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} /
+        ignoring (instance, job, type)
+        (kube_resourcequota{type="hard", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} > 0) > 1
```
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:32]: @amisstea is this ever going to be higher than one?

File: test/promql/tests/test_quota_exceeded.yaml
Comment (yftacherzog) [inline test/promql/tests/test_quota_exceeded.yaml:1]: To make the tests more realistic, I suggest also adding a resource type. You could then also include that resource in the alert message. Something like this:

```
      - series: 'kube_resourcequota{type="hard", namespace="not-reaching-limits", resourcequota="some-quota", resource="some-resource"}'
```

Comment (yftacherzog) [reply test/promql/tests/test_quota_exceeded.yaml:1]: I don't have an example in front of me, but see which labels you're getting from `kube_resourcequota` and see if including any of them on the alert rule message will be helpful for the receiver of alert.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 2]
```diff
@@ -25,3 +25,14 @@ spec:
       annotations:
         summary: "Pod {{ $labels.pod }} is crash looping"
         description: "Pod {{ $labels.namespace }}/{{ $labels.pod }} ({{ $labels.container }}) is in waiting state (reason: 'CrashLoopBackOff') for more than 15 minutes."
+    - alert: QuotaExceeded
+      expr: |
+        kube_resourcequota{type="used", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} /
+        ignoring (instance, job, type)
```
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:31]: @amisstea do you recall what was the idea behind ignoring instance and job?

**Comment by Omeramsc** [comment]
Please include a description for the PR itself

File: test/promql/tests/test_quota_exceeded.yaml
Comment (avi-biton) [reply test/promql/tests/test_quota_exceeded.yaml:1]: In continue to that, we have the `resource` label that we can use to better describe what resource quota is exceeded

File: prometheus/base/alerting/prometheus.pod_alerts.yaml [hunk 3]
```diff
@@ -25,3 +25,14 @@ spec:
...
+        (kube_resourcequota{type="hard", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} > 0) > 1
+      for: 15m
+      labels:
+        severity: warning
+      annotations:
+        summary: "Namespace {{ $labels.namespace }} exceeded its quota"
```
Comment (avi-biton) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:37]: We have the resource label that we can use to better describe what resource quota is exceeded

**Comment by Kousalya1998** [comment]
Please add a description for this PR.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 2 above)
Comment (amisstea) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:31]: I'm not quite sure. I don't see `instance` as an available label on this metric and `job` is always the same value from all samples I've seen. Perhaps the implementers of the platform rule found some edge cases where these must be ignored. I suggest we ignore only `type` for now and revisit if it becomes a problem. Let's not carry forward potential tech debt we can't explain.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 1 above)
Comment (amisstea) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:32]: Yes. I experimented by creating a `Pod` and then a `ResourceQuota` with limits and requests lower than what was defined in the `Pod`.

image

File: test/promql/tests/test_quota_exceeded.yaml
Comment (amisstea) [reply test/promql/tests/test_quota_exceeded.yaml:1]: The `container`, `resource` and `resourcequota` labels look like they would be potentially useful here.

Comment (ShakedAviv50313) [reply test/promql/tests/test_quota_exceeded.yaml:1]: I fixed it, now the tests contain also a ```resource``` and a ```resourcequota```.
The summary and the description now also contain the ```namespace```,  the ```resource``` and the ```resourcequota```

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 3 above)
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml]: @avi-biton How about now?

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 1 above)
Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:32]: Cool

Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:32]: On second thought, suppose that we consider the more standard flow in which a quota is created before the pod, will it ever alert? So maybe it should be higher or equal instead?

and also, is it possible that the quota will be 98% but resources cannot be created, but the will not fill the quota?
Suppose you have 10GB mem quota and its 9GB taken, and you try to create 2GB pods, will the quota reach 100%? Should we set the value to something smaller than 100%?

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 2 above)
Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:31]: @ShakedAviv50313 can you adjust it accordingly?
```suggestion
        ignoring (type)
```

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.pod_alerts.yaml:1]: I think this alert should be moved away from the pod alerts file, as it's not related to pod metrics.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 2 above)
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml]: Done

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:1]: Done

File: prometheus/base/alerting/prometheus.quota_alerts.yaml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+      expr: |
+        kube_resourcequota{type="used", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} /
+        ignoring (type)
+        (kube_resourcequota{type="hard", namespace!~"(.*-tenant|openshift-.*|kube-.*|default)"} > 0) > 1
+      for: 15m
+      labels:
+        severity: warning
+      annotations:
+        summary: "Resource {{ $labels.resource }} in namespace {{ $labels.namespace }} exceeded quota {{ $labels.resourcequota }}"
+        description: "Resource {{ $labels.resource }} in namespace {{ $labels.namespace }} exceeded quota {{ $labels.resourcequota }}"
```
Comment (yftacherzog) [inline prometheus/base/alerting/prometheus.quota_alerts.yaml:22]: add newline at end of file

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:1]: Need to also update kustomization.yaml so we build the new resource

File: prometheus/base/alerting/prometheus.quota_alerts.yaml (hunk 4 above)
Comment (yftacherzog) [reply prometheus/base/alerting/prometheus.quota_alerts.yaml]: I think now you have a line with some whitespaces at the end of the file. You need it to be an empty line.

File: prometheus/base/alerting/prometheus.pod_alerts.yaml
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml:1]: also done

File: prometheus/base/alerting/prometheus.quota_alerts.yaml (hunk 4 above)
Comment (ShakedAviv50313) [reply prometheus/base/alerting/prometheus.quota_alerts.yaml]: fixed it, it is now an empty line

File: prometheus/base/alerting/prometheus.pod_alerts.yaml (hunk 1 above)
Comment (amisstea) [reply prometheus/base/alerting/prometheus.pod_alerts.yaml]: Also, if we do want to fire alerts for high quota usage, I think it should be defined as a separate alert.

**Review by amisstea — APPROVED** [review]

**Pull request merged**